---
layout: page
title: Mechanics of Next Generation Sequencing Library Failure
modified: 2014-07-31T13:23:02.362000-04:00
excerpt: "Here's what I've been working on..."
---

####Why do some libraries perform far better than others during next-generation sequencing? How can we improve NGS efficiency and quality to reduce waste?

Despite the genomic data boom of the last decade and associated falling costs, sequencing studies remain expensive and time consuming. As such, failed sequencing runs -- those that produce data of insufficient quality -- represent a significant challenge and expense for genomics core facilities. 

<br>
I'm working with the McCombie Lab at [Cold Spring Harbor Laboratory](http://cshl.edu) to better characterize and quantify potential causes of PacBio sequencing failure using both computational and *in vitro* approaches. In order to quantify variables potentially implicated in NGS failure, I have developed a Python package, [SiLiCO](http://www.github.com/ethanagbaker/SiLiCO), for *in silico* read generation. [You can read more about and download SiLiCO on the SiLiCO project page.](http://silico.ethanagbaker.com) 

<br>
We recently presented our preliminary findings at AGBT 2016. See the abstract [here](/images/eagb_agbt_2016_abstract.pdf). You can also learn more about how SiLiCO works and potential uses for it in the SiLiCO paper. [[Biorxiv]] [[Bioinformatics]]
