<!doctype html>
<html>

<head>

  <title>
    
      Calculating Mean Conservation Score on Genomic Features | Ethan Alexander Garc&iacutea Baker
    
  </title>

  <meta name="viewport" content="width=device-width, initial-scale=1">

  <link rel="stylesheet" href="/css/main.css">
  <link rel="stylesheet" href="/css/syntax.css">
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=PT+Serif:400,400italic,700%7CPT+Sans:400">
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Source+Code+Pro">
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Quattrocento+Sans">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css">

  <script type="text/javascript" async
    src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-MML-AM_CHTML">
  </script>

  <script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-69741505-1', 'auto');
  ga('send', 'pageview');

</script>


</head>


<body>

  <div class="container">
    <header class="masthead">
  <h3 class="masthead-title">
    <a href="/">Ethan Alexander Garc&iacutea Baker</a>
    <small class="masthead-subtitle">computational biologist</small>
    <div class="menu">
  <div class="menu-content">
    
    <a class="element first  " href="/menu/about">about</a>
    
    <a class="element   " href="/menu/research">research</a>
    
    <a class="element  current " href="/menu/journal">journal</a>
    
    <a class="element   last" href="/menu/contact">contact</a>
    
  </div>
  <div class="social-icons">
    <a href="https://github.com/ethanagbaker"><i class="fa fa-github" aria-hidden="true" target="_blank"></i></a>
    <a href="https://twitter.com/ethanagbaker"><i class="fa fa-twitter" aria-hidden="true" target="_blank"></i></a>
    <a href="https://linkedin.com/in/ethanagbaker"><i class="fa fa-linkedin" aria-hidden="true" target="_blank"></i></a>
    <a href="mailto:ethanagbaker@gmail.com"><i class="fa fa-envelope" aria-hidden="true" target="_blank"></i></a>
  </div>
</div>

  </h3>
</header>


    <div class="post-container">
      <h1>
  Calculating Mean Conservation Score on Genomic Features
</h1>



<p>This code will calculate the mean conservation score on the features in a BED file from the USCS Genome Browser.
<br />
<br />
It requires Python and Pandas (<code class="highlighter-rouge">pip install pandas</code>).
<br />
<br />
This also requires that your feature file be a BED-like format - chromosome, start, end, and name positions must follow the BED specification, but additional columns can be appended and preserved. Data from the UCSC Conservation Track (downloaded as single base conservation) should also be converted to one BED-like file per chromosome, which can be done trivially by using the <code class="highlighter-rouge">cat -n</code> command in Terminal, adding the chromosome name (in the same style as the feature file) to each line with <code class="highlighter-rouge">awk</code>, and ensuring that all whitespace characters are tabs (<code class="highlighter-rouge">/t</code>).
<br />
<br />
A sample <code class="highlighter-rouge">featureFile</code> and <code class="highlighter-rouge">conservationFile</code> might look something like this:
<script src="https://gist.github.com/ethanagbaker/dc7bc62a413cbbc32bbf944125845afd.js"></script></p>

<p>Now, the code:</p>

<div class="language-python highlighter-rouge"><pre class="highlight"><code>  <span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">division</span>
  <span class="kn">import</span> <span class="nn">pandas</span> <span class="kn">as</span> <span class="nn">pd</span>

  <span class="n">FEATUREFILE</span> <span class="o">=</span> <span class="s">'filename'</span>
  <span class="n">CONSERVATIONFILEDIR</span> <span class="o">=</span> <span class="s">'./conservation/'</span>

  <span class="n">peakDF</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">FEATUREFILE</span><span class="p">),</span> <span class="n">sep</span> <span class="o">=</span> <span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s">'chrom'</span><span class="p">,</span><span class="s">'start'</span><span class="p">,</span><span class="s">'end'</span><span class="p">,</span><span class="s">'name'</span><span class="p">,</span><span class="s">'enrichmentVal'</span><span class="p">])</span>
  <span class="c">#Reject negative peak starts, if they exist (sometimes this can happen w/ MACS)</span>
  <span class="n">peakDF</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">peakDF</span><span class="p">[</span><span class="n">peakDF</span><span class="o">.</span><span class="n">start</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
  <span class="n">peakDF</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
  <span class="n">peakDF</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s">'index'</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
  <span class="n">peakDF</span><span class="p">[</span><span class="s">'conservation'</span><span class="p">]</span> <span class="o">=</span> <span class="mf">1.0</span>

  <span class="n">chromNames</span> <span class="o">=</span> <span class="n">peakDF</span><span class="o">.</span><span class="n">chrom</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>

  <span class="k">for</span> <span class="n">chromosome</span> <span class="ow">in</span> <span class="n">chromNames</span><span class="p">:</span>
    <span class="n">chromSubset</span> <span class="o">=</span> <span class="n">peakDF</span><span class="p">[</span><span class="n">peakDF</span><span class="o">.</span><span class="n">chrom</span> <span class="o">==</span> <span class="nb">str</span><span class="p">(</span><span class="n">chromosome</span><span class="p">)]</span>
    <span class="n">chromDF</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">CONSERVATIONFILEDIR</span><span class="p">)</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">chromosome</span><span class="p">)</span><span class="o">+</span><span class="s">'.bed'</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s">'chrom'</span><span class="p">,</span><span class="s">'start'</span><span class="p">,</span><span class="s">'end'</span><span class="p">,</span><span class="s">'conserveScore'</span><span class="p">])</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">chromSubset</span><span class="o">.</span><span class="n">index</span><span class="p">)):</span>
	    <span class="n">x</span> <span class="o">=</span> <span class="n">chromDF</span><span class="p">[</span><span class="n">chromDF</span><span class="o">.</span><span class="n">start</span> <span class="o">&gt;=</span> <span class="n">chromSubset</span><span class="p">[</span><span class="s">'start'</span><span class="p">][</span><span class="n">chromSubset</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">i</span><span class="p">]]]</span>
	    <span class="n">featureSubset</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">start</span> <span class="o">&lt;</span> <span class="n">chromSubset</span><span class="p">[</span><span class="s">'end'</span><span class="p">][</span><span class="n">chromSubset</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">i</span><span class="p">]]]</span>
	    <span class="n">x</span><span class="o">=</span><span class="bp">None</span>
	    <span class="n">featureConservation</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">featureSubset</span><span class="o">.</span><span class="n">conserveScore</span><span class="p">)</span><span class="o">/</span><span class="p">(</span><span class="n">chromSubset</span><span class="p">[</span><span class="s">'end'</span><span class="p">][</span><span class="n">chromSubset</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span><span class="o">-</span><span class="n">chromSubset</span><span class="p">[</span><span class="s">'start'</span><span class="p">][</span><span class="n">chromSubset</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">i</span><span class="p">]]))</span>
	    <span class="n">peakDF</span><span class="o">.</span><span class="n">set_value</span><span class="p">(</span><span class="n">chromSubset</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="s">'conservation'</span><span class="p">,</span><span class="n">featureConservation</span><span class="p">)</span>
	    <span class="n">featureSubset</span><span class="o">=</span><span class="bp">None</span>
  <span class="n">peakDF</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s">"featureConservation.td"</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">)</span>
</code></pre>
</div>

<p>This will write a tab-delimited, BED-like file of calculated mean conservation values on each feature in <code class="highlighter-rouge">featureFile</code>. It should look something like this:
<script src="https://gist.github.com/ethanagbaker/1749196f79a23687fe4fa6c6f025c0e8.js"></script></p>


<span class="post-date">
  Written on
  
  June
  27th,
  2016
  by Ethan Alexander García Baker
</span>
<div class="post-date">Feel free to share!</div>
  <div class="sharing-icons">
    <a href="https://twitter.com/intent/tweet?text=Calculating Mean Conservation Score on Genomic Features&amp;url=/journal/calculate-feature-conserv/&amp;via=ethanagbaker" target="_blank"><i class="fa fa-twitter" aria-hidden="true"></i></a>
    <a href="https://www.facebook.com/sharer/sharer.php?u=/journal/calculate-feature-conserv/&amp;title=Calculating Mean Conservation Score on Genomic Features" target="_blank"><i class="fa fa-facebook" aria-hidden="true"></i></a>
    <a href="https://plus.google.com/share?url=/journal/calculate-feature-conserv/" target="_blank"><i class="fa fa-google-plus" aria-hidden="true"></i></a>
  </div>
</div>

<div class="related">
  <h1 >You may also enjoy:</h1>
  
  <ul class="related-posts">
    
      
        
          <li>
            <h3>
              <a href="/journal/silico-preprint-available/">
                SiLiCO: A Simulator of Long Read Sequencing in PacBio and Oxford Nanopore
                <!--<img src="http://localhost:4000/images/silico.jpg">-->
                <!--<small>October 2, 2016</small>-->
              </a>
            </h3>
          </li>
          
        
      
        
        
      
    
      
        
        
      
        
        
      
    
  </ul>
</div>



  <section class="disqus">
  <div id="disqus_thread"></div>
  <script type="text/javascript">
    var disqus_shortname = 'ethanagbaker';
    var disqus_developer = 0;
    (function() {
      var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
      dsq.src = 'https://' + disqus_shortname + '.disqus.com/embed.js';
      (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
  </script>
  <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
  <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
</section>

    </div>

    <footer class="footer">
  <a href="https://github.com/ethanagbaker"><i class="fa fa-github" aria-hidden="true" target="_blank"></i></a>
  <a href="https://twitter.com/ethanagbaker"><i class="fa fa-twitter" aria-hidden="true" target="_blank"></i></a>
  <a href="https://linkedin.com/in/ethanagbaker"><i class="fa fa-linkedin" aria-hidden="true" target="_blank"></i></a>
  <a href="mailto:ethanagbaker@gmail.com"><i class="fa fa-envelope" aria-hidden="true" target="_blank"></i></a>
  <div class="post-date"><a href="/menu/about.html">Ethan Alexander Garc&iacutea Baker | computational biologist by Ethan Alexander García Baker</a></div>
</footer>


  </div>

</body>
</html>
